src/server/pfs/server/master.go

```diff
@@ -34,9 +42,122 @@ func (d *driver) master(ctx context.Context) {
 			gc := chunk.NewGC(d.storage.ChunkStorage())
 			return gc.RunForever(ctx)
 		})
+		eg.Go(func() error {
+			return d.finishCommits(ctx)
+		})
 		return eg.Wait()
 	}, backoff.NewInfiniteBackOff(), func(err error, _ time.Duration) error {
 		log.Errorf("error in pfs master: %v", err)
 		return nil
 	})
 }
+
+func (d *driver) finishCommits(ctx context.Context) error {
+	return d.commits.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
+		if ev.Type == watch.EventError {
+			return ev.Err
+		}
+		var key string
+		commitInfo := &pfs.CommitInfo{}
+		if err := ev.Unmarshal(&key, commitInfo); err != nil {
+			return err
+		}
+		if commitInfo.Finishing == nil || commitInfo.Finished != nil {
+			return nil
+		}
+		return backoff.RetryUntilCancel(ctx, func() error {
+			id, err := d.getFileSet(ctx, commitInfo.Commit)
+			if err != nil {
+				if pfsserver.IsCommitNotFoundErr(err) {
+					return nil
+				}
+				return err
+			}
+			// Compact the commit.
+			compactedId, err := d.compactor.Compact(ctx, []fileset.ID{*id}, defaultTTL)
+			if err != nil {
+				return err
+			}
+			if err := d.commitStore.SetTotalFileSet(ctx, commitInfo.Commit, *compactedId); err != nil {
+				return err
+			}
+			// Validate the commit.
```

Comment:
I really think this should be split out as a separate method/function.

Reply:
Especially since we will want a well defined begging and end in the code to point engineers to when they ask "what do we do in validation"

Reply:
I didn't for now because we would need to disambiguate errors reading the file set from errors associated with validation. The ergonomics for errored commits need work.